### ISSUE TITLE
Iteration runtime and GPU memory usage increase with iteration count

### ISSUE THREAD

**Issue by mr-Mojo** [mitsuba-renderer/mitsuba3#477]
I found that when using Mitsuba3 with DrJit, the iteration time increases with every call to `mi.render()`. Moreover, the GPU usage increases by a few megabyte every second (Iter0: 3.36GB, Iter500: 3.79GB). 

This should be easily reproducible with the official [pose estimation example](https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/object_pose_estimation.html), which I extended to measure the iteration time and appended below. 

I am using Mitsuba 3.1.0, but also tried with 3.0.2, with the same results. My system configuration: 
```
System information:

  OS: Ubuntu 20.04.4 LTS
  CPU: 12th Gen Intel(R) Core(TM) i9-12900K
  GPU: NVIDIA GeForce RTX 3090
  Python: 3.9.7 (default, Sep 16 2021, 13:09:58) [GCC 7.5.0]
  NVidia driver: 520.61.05
  CUDA: 11.7.64
  LLVM: 12.0.0

  Dr.Jit: 0.3.0
  Mitsuba: 3.1.0
     Is custom build? False
     Compiled with: GNU 10.2.1
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb
```

MWE to reproduce: 

```python
import time
import drjit as dr
import mitsuba as mi
from matplotlib import pyplot as plt
from mitsuba.scalar_rgb import Transform4f as T

mi.set_variant('cuda_ad_rgb')
integrator = {'type': 'direct_reparam'}

scene = mi.load_dict({
    'type': 'scene',
    'integrator': integrator,
    'sensor':  {
        'type': 'perspective',
        'to_world': T.look_at(
                        origin=(0, 0, 2),
                        target=(0, 0, 0),
                        up=(0, 1, 0)
                    ),
        'fov': 60,
        'film': {
            'type': 'hdrfilm',
            'width': 64,
            'height': 64,
            'rfilter': { 'type': 'gaussian' },
            'sample_border': True
        },
    },
    'wall': {
        'type': 'obj',
        'filename': '../scenes/meshes/rectangle.obj',
        'to_world': T.translate([0, 0, -2]).scale(2.0),
        'face_normals': True,
        'bsdf': {
            'type': 'diffuse',
            'reflectance': { 'type': 'rgb', 'value': (0.5, 0.5, 0.5) },
        }
    },
    'bunny': {
        'type': 'ply',
        'filename': '../scenes/meshes/bunny.ply',
        'to_world': T.scale(6.5),
        'bsdf': {
            'type': 'diffuse',
            'reflectance': { 'type': 'rgb', 'value': (0.3, 0.3, 0.75) },
        },
    },
    'light': {
        'type': 'obj',
        'filename': '../scenes/meshes/sphere.obj',
        'emitter': {
            'type': 'area',
            'radiance': {'type': 'rgb', 'value': [1e3, 1e3, 1e3]}
        },
        'to_world': T.translate([2.5, 2.5, 7.0]).scale(0.25)
    }
})

img_ref = mi.render(scene, seed=0, spp=1024)

mi.util.convert_to_bitmap(img_ref)
params = mi.traverse(scene)
initial_vertex_positions = dr.unravel(mi.Point3f, params['bunny.vertex_positions'])
opt = mi.ad.Adam(lr=0.025)
opt['angle'] = mi.Float(0.25)
opt['trans'] = mi.Point2f(0.1, -0.25)


def apply_transformation(params, opt):
    opt['trans'] = dr.clamp(opt['trans'], -0.5, 0.5)
    opt['angle'] = dr.clamp(opt['angle'], -0.5, 0.5)

    trafo = mi.Transform4f.translate([opt['trans'].x, opt['trans'].y, 0.0]).rotate([0, 1, 0], opt['angle'] * 100.0)

    params['bunny.vertex_positions'] = dr.ravel(trafo @ initial_vertex_positions)
    params.update()


apply_transformation(params, opt)
img_init = mi.render(scene, seed=0, spp=1024)
mi.util.convert_to_bitmap(img_init)
iteration_count = 500
spp = 16
loss_hist, times = [], []
for it in range(iteration_count):
    start = time.time()

    # has no effect
    # scene.sensors()[0].sampler().seed(it, 128)

    # Apply the mesh transformation
    apply_transformation(params, opt)

    # Perform a differentiable rendering
    img = mi.render(scene, params, seed=it, spp=spp)

    # Evaluate the objective function
    loss = dr.sum(dr.sqr(img - img_ref)) / len(img)

    # Backpropagate through the rendering process
    dr.backward(loss)

    # Optimizer: take a gradient descent step
    opt.step()
    loss_hist.append(loss)
    times.append(time.time()-start)
    print(f"Iteration {it}: error={loss[0]} - iter time: {times[-1]}")


plt.plot(times[10:])        # exclude the first few entries to ignore compilation times
plt.xlabel('Iterations')
plt.ylabel('Time per Iteration')
plt.show()

```

And the output it produces: 
![image](https://user-images.githubusercontent.com/25490579/210815744-7b350aae-743d-4060-813d-e8020af27bfe.png)

I saw the fixes suggested in #67 but am not quite sure how to use them with dr.eval(). I tried a few different variants, but none worked. I also tried adding the line suggested in #3, but that didn't help either. I also uncommented all lines within the loop but the call to `mi.render()`, and the problem persisted, so I'm guessing that it really has to do with the sampler/other internals. Help is greatly appreciated!

**Comment by njroussel** [comment]
Hi @mr-Mojo 

I ran your code as is and was unable to reproduce this behavior on the latest version (`master` build). To be honest, we've accumulated quite a few fixes that have not been released in a PyPi version yet and I can't recall exactly what has changed  since :sweat_smile:

We should be releasing a new version tomorrow (06.01.2023), hopefully it fixes your issue. I'll ping you on this issue once the new version is released.
